### ISSUE TITLE
Logstash output randomly loses records based on spool_size

### ISSUE THREAD

**Issue by roji** [elastic/filebeat#226]
I'm trying to set up filebeat rc1 to ship logs to logstash and am running into a strange issue. My config is set up to harvest exactly one file, and I clean the state before starting to make filebeat start at the beginning of the file. With the default spool_size (1024), I can see that only around 10 out of 600 events are actually shipped to logstash and inserted into elasticsearch. My filebeat config also has the file output configured, and I can see all events in the file, pointing to a problem with the filebeat-logstash interaction.

As I reduce the spool_size parameter, the number of lost records is reduced, and at spool_size == 5 everything is shipped correctly. Note that idle_timeout is configured to 5s - according to the docs this should make filebeat ship all records eventually, but this doesn't happen.

My config:

```yml
filebeat:
  prospectors:
    -
      paths:
        - /tmp/some_file.log

      fields:
        instance: some_id
      fields_under_root: true
      ignore_older: 24h
      document_type: some_type

  spool_size: 100
  idle_timeout: 5s
  registry_file: /var/lib/filebeat/registry
  config_dir:

output:
  logstash:
    enabled: true
    hosts: ["elk.eaglelabs.local:5044"]

  file:
    enabled: true
    path: "/tmp/filebeat"
    filename: filebeat

shipper:
```

**Comment by ph** [comment]
Any errors in the logs from both filebeat/logstash?

**Comment by bertobert** [comment]
I am currently struggling with the exact same issue. If you set the output to elasticsearch all events are delivered correctly as well.

**Comment by roji** [comment]
@ph, no, no errors at all, everything is reporting success...

**Comment by ph** [comment]
@urso this might be a problem with the partial ack, if it work with a window of 5, with that specific size of window the input beats won't do any partial ack for a window of that size. I need to do some test on my side.

**Comment by ph** [comment]
@roji I can reproduce the issue you are describing with filebeat rc1, With a simple script that generate a log with 10000 events and use filebeat to send it to logstash, I only get **~1600** events!!

This log is taken before sending to the logstash queue inside the plugin.
https://gist.github.com/b142b7de7ff3242fc098

code used to generate the log:
https://gist.github.com/ebedb01464d11d768ea2

filebeat config: https://gist.github.com/49b33ea4f50bc10e9ab4

I will go up in the library to see if the server is incorrectly handling the ack or not receiving everything.

**Comment by roji** [comment]
@ph great to hear you could reproduce this, looking forward to the fix!

**Comment by ruflin** [comment]
@ph Let me know in case I can help here form the Filebeat side.

**Comment by ph** [comment]
This is what I see from in the library.


```
SEQUENCE: 1 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1, "message"=>"events 0", "offset"=>0, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 2 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>2, "message"=>"events 1", "offset"=>9, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 2 ACK
SEQUENCE: 3 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>3, "message"=>"events 2", "offset"=>18, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 4 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>4, "message"=>"events 3", "offset"=>27, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 4 ACK
SEQUENCE: 5 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>5, "message"=>"events 4", "offset"=>36, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 6 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>6, "message"=>"events 5", "offset"=>45, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 6 ACK
SEQUENCE: 7 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>7, "message"=>"events 6", "offset"=>54, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 8 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>8, "message"=>"events 7", "offset"=>63, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 8 ACK
SEQUENCE: 9 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>9, "message"=>"events 8", "offset"=>72, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 10 WINDOW SIZE: 10 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.152Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>10, "message"=>"events 9", "offset"=>81, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 10 ACK
SEQUENCE: 1 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1025, "message"=>"events 1024", "offset"=>11178, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 2 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1026, "message"=>"events 1025", "offset"=>11190, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 3 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1027, "message"=>"events 1026", "offset"=>11202, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 3 ACK
SEQUENCE: 4 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1028, "message"=>"events 1027", "offset"=>11214, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 5 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1029, "message"=>"events 1028", "offset"=>11226, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 6 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1030, "message"=>"events 1029", "offset"=>11238, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 6 ACK
SEQUENCE: 7 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1031, "message"=>"events 1030", "offset"=>11250, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 8 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1032, "message"=>"events 1031", "offset"=>11262, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 9 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1033, "message"=>"events 1032", "offset"=>11274, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
SEQUENCE: 9 ACK
SEQUENCE: 10 WINDOW SIZE: 15 DATA SINGLE, content: {"@metadata"=>{"beat"=>"filebeat", "type"=>"log"}, "@timestamp"=>"2015-11-12T20:40:05.154Z", "count"=>1, "fields"=>nil, "fileinfo"=>{}, "input_type"=>"log", "line"=>1034, "message"=>"events 1033", "offset"=>11286, "shipper"=>"sashimi", "source"=>"/tmp/sample.log", "type"=>"log"}
```

This is a bit strange the first window is correct, since it contains the correctly ordered events, the inputs ack on 20% of the events and at the end of the window.

Then it received a second windows of 15 elements that start on the line 1024?

---
complete log for the 10 000 events.
https://gist.github.com/1dbd1c085584135b36e6

**Comment by ph** [comment]
@ruflin yes lets sync on that :D

**Comment by ph** [comment]
Ooops forgot to include the modified server.rb file to output that trace.
https://gist.github.com/6646166ff4fafaeb4877

**Comment by driskell** [comment]
https://github.com/elastic/libbeat/blob/master/outputs/logstash/client.go#L77

Excess events are dropped if there's more than the window size I think.

**Comment by ruflin** [comment]
@driskell Exactly. Currently trying to find a fix.

**Comment by driskell** [comment]
Would just returning the excess work? On fail it does that with the failed events, so I guess they get redistributed etc

**Comment by ruflin** [comment]
Seems to work as expected now that I return the events which were not sent yet. Now lets add some test for that :-)

**Comment by urso** [comment]
Thanks for all the help. PR for fix is ready in libbeat. Unfortunately we have had a [small bug](https://github.com/elastic/libbeat/pull/296/files#diff-b9be9406157b4d8fdc8ec4fd7e25e3eaR370) in our integration test which was supposed to detect this kind of problem early in development :(

**Comment by roji** [comment]
Great to hear, thanks for the quick fix. Will be looking forward to rc2 or to the release.
